--- a/net/minecraft/client/gui/Font.java
+++ b/net/minecraft/client/gui/Font.java
@@ -33,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public class Font {
+public class Font implements net.neoforged.neoforge.client.extensions.IFontExtension {
     private static final float EFFECT_DEPTH = 0.01F;
     private static final float OVER_EFFECT_DEPTH = 0.01F;
     private static final float UNDER_EFFECT_DEPTH = -0.01F;
@@ -42,6 +_,8 @@
     private final RandomSource random = RandomSource.create();
     private final Font.Provider provider;
     private final StringSplitter splitter;
+    /** Neo: enables linear filtering on text */
+    public boolean enableTextTextureLinearFiltering = false;
 
     public Font(Font.Provider provider) {
         this.provider = provider;
@@ -216,6 +_,8 @@
     public StringSplitter getSplitter() {
         return this.splitter;
     }
+
+    @Override public Font self() { return this; }
 
     @OnlyIn(Dist.CLIENT)
     public static enum DisplayMode {
