--- a/net/minecraft/client/resources/model/ModelBaker.java
+++ b/net/minecraft/client/resources/model/ModelBaker.java
@@ -9,7 +_,7 @@
 import org.joml.Vector3fc;
 
 @OnlyIn(Dist.CLIENT)
-public interface ModelBaker {
+public interface ModelBaker extends net.neoforged.neoforge.client.extensions.ModelBakerExtension {
     ResolvedModel getModel(Identifier location);
 
     BlockStateModelPart missingBlockModelPart();
@@ -21,7 +_,7 @@
     <T> T compute(ModelBaker.SharedOperationKey<T> key);
 
     @OnlyIn(Dist.CLIENT)
-    public interface Interner {
+    public interface Interner extends net.neoforged.neoforge.client.extensions.ModelBakerInternerExtension {
         Vector3fc vector(Vector3fc vector);
 
         BakedQuad.MaterialInfo materialInfo(BakedQuad.MaterialInfo material);
