--- a/net/minecraft/world/level/block/TallGrassBlock.java
+++ b/net/minecraft/world/level/block/TallGrassBlock.java
@@ -12,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class TallGrassBlock extends VegetationBlock implements BonemealableBlock {
+public class TallGrassBlock extends VegetationBlock implements BonemealableBlock, net.neoforged.neoforge.common.IShearable {
     public static final MapCodec<TallGrassBlock> CODEC = simpleCodec(TallGrassBlock::new);
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 13.0);
 
