--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -16,7 +_,7 @@
 import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
-public final class Transformation {
+public final class Transformation implements net.neoforged.neoforge.common.extensions.ITransformationExtension {
     private static final Vector3fc ZERO_TRANSLATION = new Vector3f();
     private static final Vector3fc UNIT_SCALE = new Vector3f(1.0F, 1.0F, 1.0F);
     private static final Quaternionfc ZERO_ROTATION = new Quaternionf();
@@ -154,6 +_,19 @@
     @Override
     public int hashCode() {
         return Objects.hash(this.matrix);
+    }
+
+    private Matrix3f normalTransform = null;
+    public org.joml.Matrix3fc getNormalMatrix() {
+        checkNormalTransform();
+        return normalTransform;
+    }
+    private void checkNormalTransform() {
+        if (normalTransform == null) {
+            normalTransform = new Matrix3f(matrix);
+            normalTransform.invert();
+            normalTransform.transpose();
+        }
     }
 
     public Transformation slerp(Transformation that, float progress) {
