--- a/net/minecraft/data/loot/LootTableSubProvider.java
+++ b/net/minecraft/data/loot/LootTableSubProvider.java
@@ -5,6 +_,6 @@
 import net.minecraft.world.level.storage.loot.LootTable;
 
 @FunctionalInterface
-public interface LootTableSubProvider {
+public interface LootTableSubProvider extends net.neoforged.neoforge.common.extensions.LootTableSubProviderExtension {
     void generate(BiConsumer<ResourceKey<LootTable>, LootTable.Builder> output);
 }
