--- a/net/minecraft/core/component/DataComponentHolder.java
+++ b/net/minecraft/core/component/DataComponentHolder.java
@@ -3,7 +_,7 @@
 import java.util.stream.Stream;
 import org.jspecify.annotations.Nullable;
 
-public interface DataComponentHolder extends DataComponentGetter {
+public interface DataComponentHolder extends DataComponentGetter, net.neoforged.neoforge.common.extensions.IDataComponentHolderExtension {
     DataComponentMap getComponents();
 
     @Override
