--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
@@ -11,9 +_,11 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public interface BlockStateModelPart {
+public interface BlockStateModelPart extends net.neoforged.neoforge.client.extensions.BlockStateModelPartExtension {
     List<BakedQuad> getQuads(@Nullable Direction direction);
 
+    /** @deprecated Neo: Use {@link #ambientOcclusion()} instead. */
+    @Deprecated
     boolean useAmbientOcclusion();
 
     Material.Baked particleMaterial();
