--- a/net/minecraft/client/renderer/block/dispatch/ModelState.java
+++ b/net/minecraft/client/renderer/block/dispatch/ModelState.java
@@ -8,7 +_,7 @@
 import org.joml.Matrix4fc;
 
 @OnlyIn(Dist.CLIENT)
-public interface ModelState {
+public interface ModelState extends net.neoforged.neoforge.client.extensions.ModelStateExtension {
     Matrix4fc NO_TRANSFORM = new Matrix4f();
 
     default Transformation transformation() {
