--- a/net/minecraft/client/renderer/SubmitNodeStorage.java
+++ b/net/minecraft/client/renderer/SubmitNodeStorage.java
@@ -29,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public class SubmitNodeStorage implements SubmitNodeCollector {
+public class SubmitNodeStorage implements SubmitNodeCollector, net.neoforged.neoforge.client.extensions.SubmitNodeStorageExtension {
     private final Int2ObjectAVLTreeMap<SubmitNodeCollection> submitsPerOrder = new Int2ObjectAVLTreeMap<>();
 
     public SubmitNodeCollection order(int order) {
